fl4health/privacy_mechanisms/dpBinomial.py

```diff
@@ -0,0 +1,38 @@
+from typing import List
+
+import torch
+
+
+# TODO why doesn't mechanism result in output type List[int]
+def BinomialMechanism(query_vector: List[int], N: int, p: float, j: int) -> List[int]:
+    """
+    Privatization of query_vector via additive binomial noise.
+
+    Args:
+        query_vector (List[int]): A vector with integer valued components.
+        N (int): The number N of Bernoulli trials which determines the distribution Bin(N, p).
+        p (float): The success probability p which determines the distribution Bin(N, p).
+        j (int): Determines the quantization scale s = 1/j.
+
+    Returns:
+        List[int]: Privatized query vector via the scaler binomial mechanism applied to each component of query_vector
+
+    Reference
+        cpSGD: Communication-efficient and differentially-private distributed SGD
+        https://proceedings.neurips.cc/paper_files/paper/2018/file/21ce689121e39821d07d04faab328370-Paper.pdf
+    """
```

Comment:
@masi-sh: Please double check my understanding here

Based on the reference above, the release mechanism M is actually a map from f(D) -> R^d. So even though you get a discrete vector in the form of f(D), M(f(D)) is a real valued vector. So the annotation List[int] for the output isn't correct. 

I'm also a little confused as to why Z (which is the variable `binomial_samples` here) is being distributed as Bin(d, p), where d is the dimension of the query vector rather than Bin(N, p) here. My understanding of `torch.distributions.Binomial` is that the first argument is the number of trials to run for each p value in the vector passed to the second argument.

Reply:
Good catch, the correct parameter should be N.

You are absolutely right about your first point, I was also talking with @masi-sh about this. In the formula both $p$ and $s$ would make the output not an integer.

Reply:
I learned that github can render latex. Very cool.